### PR TITLE
Arm Action Server

### PR DESCRIPTION
**Pull request by AdamuKaapan**
# Description
Adds the Arm Action Server, which serves the purpose of taking in target joint angles for the arm, interfacing with the `move_group` node provided by MoveIt!, and then returning the result of the execution. Also provides a test client (which will be described later).

# TODO
The one remaining part of this component is determining how the move_group "move()" method returns errors. In this case, this would mean that a plan was successfully created, but the motors themselves failed in executing the motion (for example, the scoop got stuck in the ground). I have yet to find any documentation on this, and a quick test produced no results: it could be an exception, return value, or some other method that I don't know about.

# Testing
To test this there are two launch files that must be used in tandem. First, start `roslaunch tfr_control test_armserver.launch`. This starts all of the MoveIt! nodes, as well as the ArmActionServer and Rviz to view the robot state. Second, start `roslaunch tfr_control test_armclient.launch`. This should open a logger window to view ROS console output, as roslaunch for some reason doesn't show logger output in the console and it makes me sad.

## Expected Results
1. You should see the robot appear in Rviz upon running the first launch file.
2. Once you open the second launch file, you should see the arm begin to move in an approximated digging motion, and the console window should begin to show results.
3. After the arm dumps into the bin, it will attempt to make an invalid move (where it would clip into the bin), upon which there should be an error in the console window and the arm should stay where it is.
4. The test client should exit.

### CODE REVIEW

File: src/tfr_control/CMakeLists.txt
Comment (collinsc) [inline src/tfr_control/CMakeLists.txt:47]: Not a required change, but to satisfy my curiosity, what messages does this generate and why?

File: src/tfr_control/config/test_client_positions.yaml
Comment (collinsc) [inline src/tfr_control/config/test_client_positions.yaml]: This file is fine, but I have a naming question about the folder it's stored in. 

We never established any convention or standardization for this, but this folder is called config.

Here is the naming convention I have been using:
- launch
  - launch files
- src
  - c++/python code
- include
  - headers
- params
  - launch parameters for really long and complicated nodes
- test
  - google test files

we never really picked a standard for this, but we should talk about it monday.

Maybe for files like this, a data folder?
I don't know, let me know what makes sense to you.

File: src/tfr_control/src/arm_action_server.cpp
Comment (collinsc) [inline src/tfr_control/src/arm_action_server.cpp]: Is this comment talking about the arm action or example action, it's a little unclear.

File: src/tfr_control/src/arm_action_server.cpp
Comment (collinsc) [inline src/tfr_control/src/arm_action_server.cpp]: Also you do this below
`#include <tfr_msgs/ArmMoveAction.h>`

File: src/tfr_control/src/arm_action_server.cpp
Comment (collinsc) [inline src/tfr_control/src/arm_action_server.cpp]: It's not wrong, but in recent years, the `using` keyword, has started coming more into favor. It provides better template interoperability, and interoperability with some language constructs. 

However it basically accomplishes the same task of `typedef`, it's not a required change, I'll leave it up to you. 

Here is a little light reading for you:
https://stackoverflow.com/questions/10747810/what-is-the-difference-between-typedef-and-using-in-c11

File: src/tfr_control/src/arm_action_server.cpp
Comment (collinsc) [inline src/tfr_control/src/arm_action_server.cpp]: So you store a raw pointer as a global variable, that is used in an action server with an execute callback. 

Execute callbacks get started threaded by default so you are entering a potential shared memory "danger zone", you might not have made any mistakes, but the people maintaining your code might not be as dilligent. Global variables and threading don't play nicely. 

Probably what would make sense is to have a helper object or class that can hold
```
moveit::planning_interface::MoveGroupInterface *move_group;
const robot_state::JointModelGroup *joint_model_group;
```
as `private` member variables.


Here is how I would write it, but you do you. My one requirement is avoid the unscoped publicly accessible global variables in a threaded environment.

```
class MyClass
{
public:
  MyClass(ros::NodeHandle &n) :
      move_group{"arm_end"},
      joint_model_group{move_group.getCurrentState()->getJointModelGroup("arm_end")},
      server{n, "move_arm", boost::bind(&MyClass::execute, this, _1), false}
  { server.start(); }
  
...
private:
...
void execute(const tfr_msgs::ArmMoveGoalConstPtr& goal)
{
...
server.setSucceeded(foo);
}
...
  moveit::planning_interface::MoveGroupInterface &move_group;
  const robot_state::JointModelGroup &joint_model_group;
  actionlib::SimpleActionServer<tfr_msgs::ArmMoveAction>  &server;
}
}

int main(...)
{
...
ros::NodeHandle n;
ros::AsyncSpinner spinner(1);
spinner.start();
...
MyObj obj(n);
while (ros::ok()) ;
return 0;
}
```

File: src/tfr_control/src/test_client.cpp
Comment (collinsc) [inline src/tfr_control/src/test_client.cpp]: This is a perfect situation for a `ranged for` or `foreach`:
- http://en.cppreference.com/w/cpp/language/range-for
- http://www.cplusplus.com/reference/algorithm/for_each/

But "do something for every element in a collection" is a place this stuff can cut down of the boilerplate.

However this is test code so I don't really care and the change is optional.

File: src/tfr_control/src/test_client.cpp
Comment (collinsc) [inline src/tfr_control/src/test_client.cpp]: In the future if these are parameters for just that node, we should prefer private namespace scope, to avoid namespace pollution.

Comment (collinsc) [reply src/tfr_control/src/test_client.cpp]: http://wiki.ros.org/roscpp/Overview/Parameter%20Server
Read about how to access and use private parameters.

Comment (collinsc) [reply src/tfr_control/src/test_client.cpp]: However again, test code so not a required change

File: src/tfr_control/src/test_client.cpp
Comment (collinsc) [inline src/tfr_control/src/test_client.cpp]: Does everything that interacts with the moveit library need to be spun asynchronously?

**Review by collinsc — CHANGES_REQUESTED** [review]
I asked a few questions, made a bunch of optional suggestions, and had one required change. There is a section where you need to remove some future threading risks.

Other than that looks good! It was really fun to get to watch it move!

I found some strange logs and errors, but those are in another comment.

**Comment by collinsc** [comment]
@AdamuKaapan Side Note: besides the expected warnings described in your testing report. I get some extra warining messages.
roslaunch tfr_control test_armclient.launch
image

example:
```
Dropping first 1 trajectory point(s) out of 33, as they occur before the current time.
First valid point will be reached in 0.545s.
```
Pull up rqt console, and view all transmitted messages, while running. Let me know if you see these. It it's just me whatever that's fine.

also this message was displayed to `stdout` when I ran `roslaunch tfr_control test_armserver.launch`
```
[ERROR] [1518427054.745425142]: Got a callback on a goalHandle that we're not tracking.                 This is an internal SimpleActionClient/ActionClient bug.                 This could also be a GoalID collision

```

File: src/tfr_control/CMakeLists.txt
Comment (AdamuKaapan) [reply src/tfr_control/CMakeLists.txt:47]: This currently uses messages from a long time ago, so I'd be fine with removing the whole executable itself. I just went through and added this to all the executable builds to ensure that they don't build without messages in case they need them (which most of them usually will because, hey, we're working in ROS).

File: src/tfr_control/config/test_client_positions.yaml
Comment (AdamuKaapan) [reply src/tfr_control/config/test_client_positions.yaml]: Yeah, I agree, this would be a good file to move. `data` seems like a good name for this case, I agree

File: src/tfr_control/src/arm_action_server.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/arm_action_server.cpp]: Oh, that was a copy-pasta error, I'll be sure to fix that

File: src/tfr_control/src/arm_action_server.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/arm_action_server.cpp]: Thanks, I'll fix that as well

File: src/tfr_control/src/arm_action_server.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/arm_action_server.cpp]: I mainly was just copying from the sample action server in tfr_util. For now I'll stick with what I have for consistency (although I'm completely fine with either, it makes no difference).

File: src/tfr_control/src/test_client.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/test_client.cpp]: Good note, I'll do that for the future. Might do it now just for a test/practice.

File: src/tfr_control/src/test_client.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/test_client.cpp]: I'll give it a try just to get familiar with it.

**Comment by AdamuKaapan** [comment]
@collinsc I see those errors as well, I believe that's an error with MoveIt or ros_control and not something we can control. I'll do some research to be sure.

**Comment by AdamuKaapan** [comment]
@collinsc Although that second error I'm not as sure about, I'll investigate that one more

File: src/tfr_control/src/test_client.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/test_client.cpp]: I don't think everything does (might need some more testing on that), but it seems that instantiating MoveIt needs to be.

File: src/tfr_control/CMakeLists.txt
Comment (collinsc) [reply src/tfr_control/CMakeLists.txt:47]: Well all of the message generation dependenices are in the `tfr_msgs` package. To use the messages you don't need to take a dependency on message generation, only `tfr_msgs`

File: src/tfr_control/src/arm_action_server.cpp
Comment (collinsc) [reply src/tfr_control/src/arm_action_server.cpp]: Yup it's a small thing. I ended up not using the typdef at all in my code, to each their own.

File: src/tfr_control/src/test_client.cpp
Comment (collinsc) [reply src/tfr_control/src/test_client.cpp]: `for (auto iter = position_names.begin(); iter != position_names.end(); iter++)`
can become:
`for (auto &position : position_names)
{
....
}`

File: src/tfr_control/CMakeLists.txt
Comment (AdamuKaapan) [reply src/tfr_control/CMakeLists.txt:47]: Hmmm, interesting, I'll need to check if that works to be sure, but thanks for pointing that out.

File: src/tfr_control/src/test_client.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/test_client.cpp]: Oooh, that's perfect. I'll use that.

File: src/tfr_control/CMakeLists.txt
Comment (RyanBerge) [reply src/tfr_control/CMakeLists.txt:47]: @collinsc Supposedly this was a fix to a build problem Adam was having a few weeks ago, but that problem never quite made sense to me.

File: src/tfr_control/src/arm_action_server.cpp
Comment (RyanBerge) [reply src/tfr_control/src/arm_action_server.cpp]: I go back and forth.  `using` is a newer standard and is probably preferred, but either way is fine.

File: src/tfr_control/config/test_client_positions.yaml
Comment (RyanBerge) [reply src/tfr_control/config/test_client_positions.yaml]: `config` and `data` both seem fine to me.

File: src/tfr_control/src/arm_action_server.cpp
Comment (RyanBerge) [inline src/tfr_control/src/arm_action_server.cpp]: This is definitely optional, but when you have something as long as `moveit::planning_interface::MoveItErrorCode::SUCCESS` I think adding a `typedef` or `using` statement is perfectly reasonable.

File: src/tfr_control/src/arm_action_server.cpp
Comment (RyanBerge) [inline src/tfr_control/src/arm_action_server.cpp]: You have a comment that asks how errors are handled.  Are you asking how the `move` function handles errors, or how we want to handle them?

File: src/tfr_control/src/arm_action_server.cpp
Comment (RyanBerge) [reply src/tfr_control/src/arm_action_server.cpp]: I don't think it has to be a class, but I agree that it shouldn't be stored globally here with no shared-data protections.  I don't think there's a race condition in the code you've written, but that doesn't mean it couldn't be introduced later by accident.

**Review by RyanBerge — CHANGES_REQUESTED** [review]
I agree with pretty much everything @collinsc said, and I asked a couple of questions of my own.

File: src/tfr_control/config/test_client_positions.yaml
Comment (collinsc) [reply src/tfr_control/config/test_client_positions.yaml]: I think they both work also, but let's pick a naming convention eventually.

File: src/tfr_control/src/arm_action_server.cpp
Comment (collinsc) [reply src/tfr_control/src/arm_action_server.cpp]: Functionally the same, however `using` can help you avoid one particularly nasty template use case. See above article.

File: src/tfr_control/src/arm_action_server.cpp
Comment (collinsc) [reply src/tfr_control/src/arm_action_server.cpp]: What alternatives to a class would you suggest?

File: src/tfr_control/CMakeLists.txt
Comment (collinsc) [reply src/tfr_control/CMakeLists.txt:47]: Weird, that is bizarre. also this has been bugging me. @AdamuKaapan the `proto` prefix was something we used waaaaayyyyy back when we were porting stuff over from the prototyping code, and went out of use very shortly after the start of development. Are we using `proto_driver_station` for anything, or is that a piece of test code, or something from a previous sprint?

Comment (AdamuKaapan) [reply src/tfr_control/CMakeLists.txt:47]: `proto_driver_station` was something you added when you did the initial gazebo commit, yeah. We don't need this executable any more, we can remove it in another PR soon.

File: src/tfr_control/src/arm_action_server.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/arm_action_server.cpp]: Thanks, that's a good idea: went ahead and implemented it in an upcoming commit.

File: src/tfr_control/src/arm_action_server.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/arm_action_server.cpp]: I'm asking how the move function handles errors: I know how I want to handle them I think, it'll just be returning a failure and maybe stopping execution of the arm command.

File: src/tfr_control/src/arm_action_server.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/arm_action_server.cpp]: I think I'm going to go with a class, @collinsc showed me that pattern earlier and it does seem really clean and helps solve some other issues I have sometimes.

File: src/tfr_control/src/test_client.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/test_client.cpp]: Implemented this in the upcoming commit.

File: src/tfr_control/src/arm_action_server.cpp
Comment (RyanBerge) [reply src/tfr_control/src/arm_action_server.cpp]: @collinsc You can also create static variables inside of free functions, which are basically scoped globals, but allow for access guards.  I don't really think that's better, but it does work.

Comment (collinsc) [reply src/tfr_control/src/arm_action_server.cpp]: For the sake of my own curiosity, could you put that in pseudocode?

File: src/tfr_control/CMakeLists.txt
Comment (collinsc) [reply src/tfr_control/CMakeLists.txt:47]: Thanks, I'll make a trello card for that.

File: src/tfr_control/src/arm_action_server.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/arm_action_server.cpp]: ... took me literally 30 seconds to find the answer to this, so I'm good now. Added a quick super simple implementation of it, incoming in like 5 minutes.

**Review by collinsc — DISMISSED** [review]
Looks great! thanks for implementing that.

File: src/tfr_control/src/arm_action_server.cpp
Comment (RyanBerge) [inline src/tfr_control/src/arm_action_server.cpp]: Current coding standard says braces get their own line.

File: src/tfr_control/src/arm_action_server.cpp
Comment (RyanBerge) [inline src/tfr_control/src/arm_action_server.cpp]: See above.

**Review by RyanBerge — CHANGES_REQUESTED** [review]
Refactored class is much better.  Just a couple nitpicks about style.

File: src/tfr_control/src/arm_action_server.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/arm_action_server.cpp]: Thanks. I'm pretty sure I did this on accident in a few other places, so I'll quickly scan all of my files to be sure (already caught one in my constructor as well).

**Review by collinsc — APPROVED** [review]
redoing stale review

**Pull request merged**
